libminifi/src/core/state/nodes/ResponseNodeLoader.cpp

```diff
@@ -59,17 +59,17 @@ void ResponseNodeLoader::initializeComponentMetrics(core::ProcessGroup* root) {
     node_source->getResponseNodes(metric_vector);
     std::lock_guard<std::mutex> guard(component_metrics_mutex_);
     for (const auto& metric : metric_vector) {
-      component_metrics_[metric->getName()] = metric;
+      component_metrics_[metric->getName()].push_back(metric);
     }
   }
 }
 
-std::shared_ptr<ResponseNode> ResponseNodeLoader::getResponseNode(const std::string& clazz) const {
+std::vector<std::shared_ptr<ResponseNode>> ResponseNodeLoader::getResponseNodes(const std::string& clazz) const {
   std::shared_ptr<core::CoreComponent> ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
   if (ptr == nullptr) {
-    return getComponentMetricsNode(clazz);
+    return getComponentMetricsNodes(clazz);
   }
-  return std::dynamic_pointer_cast<ResponseNode>(ptr);
+  return {std::dynamic_pointer_cast<ResponseNode>(ptr)};
```

Comment:
is "having only non-null elements" an invariant of the return value of `getResponseNodes`, or do we check everywhere we use it for nulls?

Reply:
In my experience it's more of an invariant as it is passed through a lot of other functions, casting it in a few places as well, and in some places we check for nullptr and in other places we don't.

Reply:
in that case would a `gsl_Ensures` make sense here?

Reply:
Added a `gsl_Expects` for this in 567764bbdc4bac04b07b50337293ae732fd8580d

Reply:
Updated to `gsl_Ensures` in a771342d0fb11f56cd568aeeb427d4019b1b01c6 (you're right, seems more logical in this context)

Reply:
thinking about it some more, `gsl_Ensures` would document and well... ensure that some invariant is maintained, but that invariant should not come from or depend on `gsl_Ensures`, I think we should manually check if the dynamic cast succeeds, log some errors if not, and return an empty vector in that case (throwing could also work, not sure what would that trigger though)

Reply:
Updated in 67f0532544325afb6d51250209fa2f2440f95d24

Reply:
I think it would make more sense to document this in the return type by wrapping the shared_ptr in not_null. This would also automate the check on construction.

Reply:
As I checked that would be a bigger refactor, as this is the main source of the response nodes, that change would propagate through `MetricsBase`, `ResponseNodeLoader`, C2Client and so on everywhere where `std::shared_ptr<state::response::ResponseNode>` is used for parameters and class members. I would keep this manual check for the moment.

Reply:
Added it as an additional TODO point in https://issues.apache.org/jira/browse/MINIFICPP-1869